File: 0001_two_sum/two_sum.cpp
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

/**
 * Time: O(n)
 * Space: O(n)
 */
vector<int> two_sum(vector<int> &nums, int target)
{
    vector<int> res;
    unordered_map<int, int> map;
    for (int i = 0; i < nums.size(); i++)
    {
        int compliment = target - nums[i];
        if (map.count(compliment))
        {
            res.push_back(i);
            res.push_back(map[compliment]);
            return res;
        }
        map[nums[i]] = i;
    }
    return res;
}

File: 0001_two_sum/two_sum.java
// Time: O(n)
// Space: O(n)
class Solution {
    public int[] twoSum(int[] nums, int target) {
        HashMap<Integer, Integer> map = new HashMap<Integer, Integer>();
        for (int i = 0; i < nums.length; i++) {
            int curr = nums[i];
            int compliment = target - curr;
            if (map.containsKey(compliment))
                return new int[]{i, map.get(compliment)};
            map.put(curr, i); 
        }
        return new int[]{};
    }
}